packages/web-components/src/components/masthead/masthead-global-bar.ts

```diff
@@ -21,16 +22,70 @@ const { stablePrefix: ddsPrefix } = ddsSettings;
  */
 @customElement(`${ddsPrefix}-masthead-global-bar`)
 class DDSMastheadGlobalBar extends FocusMixin(LitElement) {
+  /**
+   * Media query on medium breakpoint.
+   */
+  private mqMedium;
+
   /**
    * The shadow slot this action bar should be in.
    */
   @property({ reflect: true })
   slot = 'profile';
 
+  /**
+   * Search bar opened flag.
+   */
+  @internalProperty()
+  private openedSearch = false;
+
+  /**
+   * Medium breakpoint flag.
+   */
+  @internalProperty()
+  private responsiveMd = false;
+
+  /**
+   * Handles toggle event from the search component.
+   *
+   * @param event The event.
+   */
+  private _handleSearchToggle(event: Event) {
+    this.openedSearch = (event as CustomEvent).detail.active;
+  }
+
+  /**
+   * Handles media query event.
+   *
+   * @param event The event.
+   */
+  private _handleMediaQuery(event: Event) {
+    console.log(event);
+    this.responsiveMd = (event as MediaQueryListEvent).matches;
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.getRootNode().addEventListener('dds-masthead-search-toggled', this._handleSearchToggle.bind(this));
+
+    this.mqMedium = window.matchMedia(`(max-width: ${breakpoints.md.width})`);
+    this.responsiveMd = this.mqMedium.matches;
+    this.mqMedium.addEventListener('change', this._handleMediaQuery.bind(this));
+  }
+
+  disconnectedCallback() {
+    this.getRootNode().removeEventListener('dds-masthead-search-toggled', this._handleSearchToggle.bind(this));
+    this.mqMedium.removeEventListener('change', this._handleMediaQuery.bind(this));
```

Comment:
`.bind()` creates new function instance upon every call, and thus the function refs here will be different from ones in `connectedCallback()`, which causes event listeners not removed.

To solve the problem, we use [Carbon core `on()` API](https://github.com/carbon-design-system/carbon/blob/v10.17.0/packages/components/src/globals/js/misc/on.js). Here's an example: https://github.com/carbon-design-system/ibm-dotcom-library/blob/master/packages/web-components/src/components/footer/footer-composite.ts#L139-L144

Reply:
Ok... This one is interesting.  Check out the `addEventListener` line again:
```
this.getRootNode().addEventListener('dds-masthead-search-toggled', this._handleSearchToggle.bind(this));
```

And checkout the same function call in the `on` utility:
```
element.addEventListener.apply(element, args);
```

Using this `apply` function there with `element` as the `addEventListener` context and the first argument in the `apply` function means that `element` becomes the `this` object in the callback used as the event listener.

But in my `addEventListener` call, the context of the `addEventListener` call is `this.getRootNode()` and the bound `this` object on the callback in the current class `DDSMastheadGlobalBar` object.  So they are different, so I can't use the `on` utility as it currently exists.

Reply:
But..... trying the properly remove the listener made me run into another curiosity.

I add the listener to `this.getRootNode()` in the `connectedCallback` and I remove the listener from `this.getRootNode()` in the `disconnectedCallback`.  But no matter how I juggled the listener callback function and it's binding, I couldn't get the listener properly removed.

Until I realized that `this.getRootNode()` returns different objects in `connectedCallback` vs. `disconnectedCallback`.  `this.getRootNode()` return the document object in `connectedCallback`.  And it returns `<div id="root-inner"></div>` in the `disconnectedCallback`.

Is this a behavior you are familiar with?  It feels unexpected to me.  I haven't been able to find any concrete references to this in my web searches.

Reply:
I meant:

```javascript
// Static `eventSearchToggle` property must be set pointing to 'dds-masthead-search-toggled'
const { eventSearchToggle } from this.constructor as typeof DDSMastheadGlobalBar;
this._hSearchToggle = on(this.getRootNode(), eventSearchToggle, this._handleSearchToggle.bind(this));
```

`this._hSearchToggle` is a tiny object with `.release()` method that you can use it in `.disconnectedCallback()`.

```javascript
disconnectedCallback() {
  if (this._hSearchToggle) {
    this._hSearchToggle = this._hSearchToggle.release(); // Removes the event listener
  }
  (Do other stuffs)
  super.disconnectedCallback();
}
```

This pattern is similar to [what React `useEffect()` returns](https://reactjs.org/docs/hooks-effect.html#example-using-hooks-1). The difference is object plus `.release()` method vs. plain function that does the release - Carbon `on()` API takes the former so that a class can implement the same interface, and is seen in [Dojo `on()` API](https://dojotoolkit.org/reference-guide/1.7/dojo/on.html).

BTW [recent commit to `master`](https://github.com/carbon-design-system/ibm-dotcom-library/pull/3626) allows an alternate approach:

```javascript
// Static `eventSearchToggle` property must be set pointing to 'dds-masthead-search-toggled'
@HostListener('parentRoot:eventSearchToggle')
// @ts-ignore: The decorator refers to this method but TS thinks this method is not referred to
private _handleSearchToggle = (event: Event) => { // Arrow function for `this` binding
  this.openedSearch = (event as CustomEvent).detail.active;
};
```

Above code attaches (and automatically detaches) the event listener with a [decorator](https://github.com/tc39/proposal-decorators).